lib/IRGen/GenCall.cpp

```diff
@@ -1417,6 +1429,18 @@ void SignatureExpansion::expandExternalSignatureTypes() {
     }
   }
 
+  if (formalIndirectResult) {
+    // If the result is a formal indirect result in SIL, that means that the
+    // Clang function has an explicit output parameter (e.g. it's a C++
+    // constructor). This means:
+    // - Don't mark it `sret`, as this should only be used for C++ return
+    //   values.
+    // - The Clang function might capture the pointer, so don't specify
+    //   `nocapture`.
+    addIndirectResultAttributes(IGM, Attrs, 0, /* allowSRet = */ false,
+                                /* noCapture = */ false);
```

Comment:
So, technically, the "i.e." doesn't hold here — we can have a semantically-indirect result that really isn't allowed to be captured this way.  That would only happen with an aggregate that wasn't a non-trivial C++ type, though: for example, a struct with an ObjC `__weak` field or a field with an address-diversified `__ptrauth` type.  So there are other reasons to maybe just let Clang handle these attributes.

Reply:
Also, general question: have you tested this on an ARM target, e.g. Apple ARM64?  Because the ARM C++ ABI actually gives constructors and destructors a return value (they return `this`), so we need to make sure we don't have unexpected mismatches.

Reply:
> Also, general question: have you tested this on an ARM target, e.g. Apple ARM64?

I think the bots run tests on a simulator so that should test on ARM64. We also have some tests that run on multiple targets no matter the local environment. (Semi-unrelated: in order to run the swift compiler for multiple targets, these tests use `-parse-as-library` which can cause odd behavior so we may want to move away from that in the future.)

> Because the ARM C++ ABI actually gives constructors and destructors a return value (they return this), so we need to make sure we don't have unexpected mismatches.

IIRC the Windows ABI also returns a pointer from constructors of non-trivial types. I had to update a few tests to accommodate that recently.

Reply:
After removing this, I discovered another reason I think keeping `addIndirectResultAttributes` here is incorrect: `addIndirectResultAttributes` will add the `noalias` attribute to the output pointer argument which I am pretty sure is not a safe assumption to make. There's no guarantee that the (user-provided) constructor will not write to a global variable, for example.